contrib/babelfishpg_tsql/sql/upgrades/babelfishpg_tsql--2.2.0--2.3.0.sql

```diff
@@ -2577,6 +2577,259 @@ AS 'babelfishpg_tsql', 'update_user_catalog_for_guest';
+  	    WHEN st.is_table_type = 1 THEN NULL
+  	    ELSE st.scale
+  	  END
+  	AS smallint) AS SCALE
+  , CAST(
+      CASE
+  	    WHEN st.is_table_type = 1 THEN NULL
+  	    ELSE sdit.num_prec_radix
+      END
+  	AS smallint) AS RADIX
+  , CAST(
+      CASE
+  	    WHEN st.is_table_type = 1 THEN 1
+  	    ELSE sdit.nullable 
+  	  END
+  	AS smallint) AS NULLABLE
+  , CAST(NULL AS sys.varchar(254)) AS REMARKS
+  , CAST(NULL AS sys.nvarchar(4000)) AS COLUMN_DEF
+  , CAST(
+      CASE
+  	    WHEN st.is_table_type = 1 THEN -153
+  	    ELSE sdit.sql_data_type 
+  	  END
+  	AS smallint) AS SQL_DATA_TYPE
+  , CAST(sdit.sql_datetime_sub AS smallint) AS SQL_DATETIME_SUB
+  , CAST(
+      CASE
+  	    WHEN st.is_table_type = 1 THEN 2147483647
+  	    ELSE NULL
+  	  END
+  	AS int) AS CHAR_OCTET_LENGTH
+  , CAST((ss.x).n AS INT) AS ORDINAL_POSITION
+  , CAST(
+      CASE
+        WHEN st.is_table_type = 1 THEN 'YES'
+        WHEN sdit.nullable = 1 THEN 'YES'
+        ELSE 'NO'
+      END
+    AS sys.varchar(254)) AS IS_NULLABLE
+  , CAST(
+  	  CASE
+  	    WHEN st.is_table_type = 1 THEN 0
+  	    ELSE sdit.ss_data_type
+  	  END
+  	AS sys.tinyint) AS SS_DATA_TYPE
+  , CAST(ss.proname AS sys.sysname) AS original_procedure_name
+  FROM pg_type t
+  INNER JOIN sys.types st ON st.user_type_id = t.oid
+  LEFT JOIN sys.spt_datatype_info_table sdit ON sdit.type_name = (SELECT name FROM sys.types WHERE user_type_id = st.system_type_id)
+  INNER JOIN
+  (
+    SELECT
+      p.oid AS p_oid,
+      p.proname,
+      p.proargnames,
+      p.proargmodes,
+      p.prokind,
+      p.proretset,
+      COALESCE(d.name, sys.db_name()) as name,
+      COALESCE(bne.orig_name, CAST(ns.nspname AS sys.sysname)) as orig_name,
+      information_schema._pg_expandarray(
+      COALESCE(p.proallargtypes,
+        CASE 
+          WHEN p.prokind = 'f' THEN (CAST(p.proargtypes AS oid[]))
+          ELSE CAST(p.proargtypes AS oid[])
+        END
+      )) AS x
+    FROM pg_proc p
+    INNER JOIN pg_catalog.pg_namespace ns ON ns.oid = p.pronamespace
+    LEFT JOIN sys.babelfish_namespace_ext bne ON ns.nspname = bne.nspname 
+    LEFT JOIN sys.databases d ON d.database_id = bne.dbid
+    WHERE (
+      p.pronamespace in (select schema_id from sys.schemas union all select oid from pg_namespace where nspname = 'sys')
+      AND (pg_has_role(p.proowner, 'USAGE') OR has_function_privilege(p.oid, 'EXECUTE'))
+      AND ( p.probin like '{%typmod_array%}'-- typmod array exists for those that are user-defined
+      		OR p.proname like 'sp\_%' -- filter out internal babelfish-specific procs
+      		OR p.proname like 'xp\_%'
+      		OR p.proname like 'dm\_%'
+      		OR p.proname like 'fn\_%')) 
+  ) ss ON t.oid = (ss.x).x
+WHERE ( ss.proargmodes[(ss.x).n] in ('i', 'o', 'b')  OR ss.proargmodes[(ss.x).n] is NULL)
+
+UNION
+-- Select return types. This needs to be unionized because inline-table functions may
+-- end up with 2 't' outputs which causes previous query to miss these items
```

Comment:
I wonder if we could UNION input and output types in one subquery and avoiding copying similar joins and case expressions

Reply:
I'll give that a shot and try it out.

Reply:
Was able to do it, please see the newest revision.